keras/layers/normalization/batch_normalization.py

```diff
@@ -282,3 +287,57 @@ def get_config(self):
             "synchronized": self.synchronized,
         }
         return {**base_config, **config}
+
+    def _moments(self, inputs, mask):
+        if mask is None:
+            return ops.moments(
+                inputs,
+                axes=self._reduction_axes,
+                synchronized=self.synchronized,
+            )
+
+        mask_weights = ops.cast(
+            mask,
+            inputs.dtype,
+        )
+        mask_weights_broadcasted = ops.expand_dims(
+            mask_weights,
+            axis=-1,
+        )
+        weighted_inputs = mask_weights_broadcasted * inputs
+
+        weighted_input_sum = ops.sum(
+            weighted_inputs,
+            self._reduction_axes,
+            keepdims=True,
+        )
+        sum_of_weights = ops.sum(
+            mask_weights_broadcasted,
+            self._reduction_axes,
+            keepdims=True,
+        )
+        mean = self._divide_no_nan(weighted_input_sum, sum_of_weights)
+
+        difference = inputs - mean
+        squared_difference = difference * difference
+        weighted_distsq = ops.sum(
+            mask_weights_broadcasted * squared_difference,
+            self._reduction_axes,
+            keepdims=True,
+        )
+        variance = self._divide_no_nan(weighted_distsq, sum_of_weights)
+
+        return ops.squeeze(mean), ops.squeeze(variance)
+
+    def _divide_no_nan(self, x1, x2):
+        original_dtype = x1.dtype
+        x2_binary = ops.cast(x2, dtype="bool")
+        x2_mask = ops.cast(x2_binary, dtype=original_dtype)
+        x1_masked = x1 * x2_mask
+
+        x2_binary_inverted = ~x2_binary
```

Comment:
Prefer using an op rather than a unary operator since it makes it easy to parse what code it is executing

Reply:
Removed.